# [LAB-03] 2. 탐색적 데이터 분석 - 연습문제

## 준비작업

### 라이브러리 참조

In [1]:
from jussam import load_data
from helpers import *
from pandas import DataFrame, concat
from IPython.display import display, Markdown

📦 연세대학교 주영아 교수가 제작한 라이브러리를 사용중입니다.
📧 Email(1): j.purplerose@yonsei.ac.kr
📧 Email(2): j.purplerose@gmail.com
📝 Website: https://juyounga.kr/
🔖 Version: 0.5.19


## 📚 킹 카운티, 점검은 끝났고 이제 변수를 골라야 한다

### 당신은 지난 단원에서 킹 카운티 주택 자료의 품질 점검을 마친 분석가입니다.

침실 33개짜리 입력 오류, 결측으로 오해하기 쉬운 0, 면적 계열의 극단적인 우측 꼬리 — 무엇이 걸리는지는 다 적어 보고했습니다. 그런데 보고서를 넘겨받은 팀장이 묻습니다. "문제가 뭔지는 알겠다. **그래서 모델에 넣을 변수는 뭐냐?**"

캘리포니아에서는 두 집단으로 나뉘는 명목형 변수가 없어서 EDA 범위표의 T검정 칸이 '미적용'으로 비어 있었습니다. 이 자료에는 워터프론트 여부가 있으니 **이번에는 그 칸이 채워집니다.**

중복 행만 정리하고 값은 원본 그대로 둡니다. 종속변수는 주택 가격입니다.

변수는 이렇게 나눠서 시작합니다. **워터프론트 여부·조망 점수·주택 상태 점수는 범주형**입니다. 워터프론트 여부는 지난 단원에서 확정한 그대로이고, 점수 두 종은 그때 연속형으로 두었지만 이번에는 집단을 나눠 가격을 비교하는 것이 목적이라 성격을 다시 정했습니다. **건축 등급 점수는 1~13으로 촘촘해 연속형으로 남깁니다.** **거래 날짜·우편번호·위경도·리모델링 연도는 제외**합니다. 우편번호는 범주형이 맞지만 지역이 70개라 집단 비교가 어렵고, 리모델링 연도는 지난 단원 결론대로 여부로 바꿔 써야 하는데 그 파생은 전처리 단계 일입니다.

지난 단원에서 "0은 없다는 뜻"이라고 함께 묶었던 조망 점수·지하 면적·리모델링 연도가 여기서 범주형·연속형·제외로 갈립니다. **성격이 같아 보여도 조치는 변수마다 다릅니다.**

> "단변량부터 다변량까지 차례로 훑고, **1순위로 채택할 변수 하나**를 근거와 함께 가져와라."

### 💻 코드 작성

#### 데이터 가져오기 (품질 점검 결과 반영)

In [2]:
origin = load_data("kc_house")

# 지난 단원의 품질 점검대로 중복 행만 정리하고, 값은 원본 그대로 사용한다.
df0 = my_qtcheck.check_duplicates(origin)

print(f"데이터 크기: {df0.shape}")
df0.head()

📚 이 데이터 세트는 시애틀이 속한 킹 카운티의 주택 매매 가격 정보를 담고 있습니다. 2014년 5월부터 2015년 5월까지 판매된 주택들이 포함되어 있습니다.

(출처: https://www.kaggle.com/datasets/harlfoxem/housesalesprediction)



    field          description
--  -------------  ------------------------------------------------
 0  date           거래 날짜
 1  price          주택 가격 (USD 달러)
 2  bedrooms       침실 수 (개수)
 3  bathrooms      욕실 수 (개수)
 4  sqft_living    거주 공간 면적 (제곱 피트)
 5  sqft_lot       대지 면적 (제곱 피트)
 6  floors         층 수 (개수)
 7  waterfront     워터프론트 여부 (이진형: 1/0)
 8  view           조망 점수 (0~4)
 9  condition      주택 상태 점수 (1~5)
10  grade          건축 등급 점수 (1~13)
11  sqft_above     지상 면적 (제곱 피트)
12  sqft_basement  지하 면적 (제곱 피트)
13  yr_built       건축 연도
14  yr_renovated   리모델링 연도
15  zipcode        우편번호
16  lat            위도
17  long           경도
18  sqft_living15  15개 인근 주택의 평균 거주 공간 면적 (제곱 피트)
19  sqft_lot15     15개 인근 주택의 평균 대지 면적 (제곱 피트)

중복된 행의 수: 0
데이터 크기: (21613, 20)


,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,condition,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
0,2014-10-13,221900.000,3,1.000,1180,5650,1.000,0,0,3,7,1180,0,1955,0,98178,47.511,-122.257,1340,5650
1,2014-12-09,538000.000,3,2.250,2570,7242,2.000,0,0,3,7,2170,400,1951,1991,98125,47.721,-122.319,1690,7639
2,2015-02-25,180000.000,2,1.000,770,10000,1.000,0,0,3,6,770,0,1933,0,98028,47.738,-122.233,2720,8062
3,2014-12-09,604000.000,4,3.000,1960,5000,1.000,0,0,5,7,1050,910,1965,0,98136,47.521,-122.393,1360,5000
4,2015-02-18,510000.000,3,2.000,1680,8080,1.000,0,0,3,8,1680,0,1987,0,98074,47.617,-122.045,1800,7503


#### 타입 변환 (범주형 확정)

In [3]:
# 지난 단원에서 확정한 범주형 2종(워터프론트 여부·우편번호)에
# 이번 단원에서 집단으로 볼 점수 2종(조망 점수·주택 상태 점수)을 더한다.
df = my_qtcheck.set_type(df0, as_category=["waterfront", "zipcode", "view", "condition"])

<class 'pandas.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 20 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   date           21613 non-null  datetime64[ns]
 1   price          21613 non-null  float64       
 2   bedrooms       21613 non-null  int64         
 3   bathrooms      21613 non-null  float64       
 4   sqft_living    21613 non-null  int64         
 5   sqft_lot       21613 non-null  int64         
 6   floors         21613 non-null  float64       
 7   waterfront     21613 non-null  category      
 8   view           21613 non-null  category      
 9   condition      21613 non-null  category      
 10  grade          21613 non-null  int64         
 11  sqft_above     21613 non-null  int64         
 12  sqft_basement  21613 non-null  int64         
 13  yr_built       21613 non-null  int64         
 14  yr_renovated   21613 non-null  int64         
 15  zipcode        21613 non-null 

#### 변수 유형 분류

In [4]:
target = "price"              # 종속변수 이름
target_is_continuous = True   # 종속변수 연속형 여부

# 분석 대상에서 제외할 변수
# -> 거래 시점·좌표, 지역이 70개라 집단 비교가 어려운 우편번호,
#    그리고 "연도가 아니라 여부"로 바꿔 써야 하는 리모델링 연도
exclude_cols = ["date", "zipcode", "lat", "long", "yr_renovated"]

nominal_cols = my_qtcheck.get_categorical_column_names(df)   # 범주형 변수
continuous_cols = my_qtcheck.get_number_column_names(df)     # 연속형 변수

if target_is_continuous:               # 종속변수가 연속형 변수라면?
    continuous_cols.remove(target)     # 연속형 변수 이름에서 종속변수 항목 제거
else:                                  # 종속변수가 범주형 변수라면?
    nominal_cols.remove(target)        # 범주형 변수 이름에서 종속변수 항목 제거

# 제외 대상으로 지정한 변수를 각 목록에서 제거한다.
continuous_cols = [c for c in continuous_cols if c not in exclude_cols]
nominal_cols = [c for c in nominal_cols if c not in exclude_cols]

print("종속변수 :", target)
print("범주형   :", nominal_cols)
print("연속형   :", continuous_cols)
print("제외     :", exclude_cols)

종속변수 : price
범주형   : ['waterfront', 'view', 'condition']
연속형   : ['bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors', 'grade', 'sqft_above', 'sqft_basement', 'yr_built', 'sqft_living15', 'sqft_lot15']
제외     : ['date', 'zipcode', 'lat', 'long', 'yr_renovated']


#### 컬럼 의미를 정리한 딕셔너리

In [5]:
column_means = {
    "price":         "주택 가격 (USD)",
    "bedrooms":      "침실 수",
    "bathrooms":     "욕실 수",
    "sqft_living":   "거주 공간 면적",
    "sqft_lot":      "대지 면적",
    "floors":        "층 수",
    "waterfront":    "워터프론트 여부",
    "view":          "조망 점수",
    "condition":     "주택 상태 점수",
    "grade":         "건축 등급 점수",
    "sqft_above":    "지상 면적",
    "sqft_basement": "지하 면적",
    "yr_built":      "건축 연도",
    "sqft_living15": "인근 15채의 평균 거주 면적",
    "sqft_lot15":    "인근 15채의 평균 대지 면적"
}

#### 단변량 분석 - 연속형 (종속변수 + 독립변수)

In [6]:
num_desc = my_qtcheck.numerical_summary(df[[target] + continuous_cols])
num_desc[['mean', '50%', 'skew', 'kurt', 'outliers_ratio', 'log_need']]

,mean,50%,skew,kurt,outliers_ratio,log_need
price,540088.142,450000.000,4.024,34.586,0.053,log
bedrooms,3.371,3.000,1.974,49.064,0.025,log1p
bathrooms,2.115,2.250,0.511,1.280,0.026,log1p
sqft_living,2079.900,1910.000,1.472,5.243,0.026,log
sqft_lot,15106.968,7618.000,13.060,285.078,0.112,log
floors,1.494,1.500,0.616,-0.485,0.000,none
grade,7.657,7.000,0.771,1.191,0.088,log
sqft_above,1788.391,1560.000,1.447,3.402,0.028,log
sqft_basement,291.509,0.000,1.578,2.716,0.023,log1p
yr_built,1971.005,1975.000,-0.470,-0.657,0.000,none


#### 단변량 분석 - 명목형

In [7]:
cat_desc = my_qtcheck.categorical_summary(df[nominal_cols])

for n in nominal_cols:
    display(Markdown(f"#### ▶︎ {n}({column_means[n]}) 분포"))

    freq = df[n].value_counts().rename("빈도").to_frame()
    freq["비율"] = df[n].value_counts(normalize=True)
    display(freq)

📊 컬럼 'waterfront'의 value_counts():


,count
waterfront,
0,21450
1,163


📊 컬럼 'view'의 value_counts():


,count
view,
0,19489
1,332
2,963
3,510
4,319


📊 컬럼 'condition'의 value_counts():


,count
condition,
1,30
2,172
3,14031
4,5679
5,1701


#### ▶︎ waterfront(워터프론트 여부) 분포

,빈도,비율
waterfront,,
0,21450,0.992
1,163,0.008


#### ▶︎ view(조망 점수) 분포

,빈도,비율
view,,
0,19489,0.902
2,963,0.045
3,510,0.024
1,332,0.015
4,319,0.015


#### ▶︎ condition(주택 상태 점수) 분포

,빈도,비율
condition,,
3,14031,0.649
4,5679,0.263
5,1701,0.079
2,172,0.008
1,30,0.001


#### 이변량 분석 - 상관분석 (연속형 → 연속형 종속변수)

In [8]:
corr_list = []

for f in continuous_cols:
    corr_list.append(my_stats.correlation(df, x=f, y=target, plot=False))

# 11개 결과를 한 표로 합쳐 상관계수 내림차순 정렬
corr_result = concat(corr_list).sort_values("coef", ascending=False)
display(corr_result)

print("[ 상관 강도 판정별 변수 개수 ]")
print(corr_result['strength'].value_counts())

,,method,coef,p-value,strength,significant,normality_x,normality_y,linearity,influential_outlier,high_skew
x,y,,,,,,,,,,
grade,price,Spearman,0.658,0.000,Moderate,True,False,False,False,True,True
sqft_living,price,Spearman,0.644,0.000,Moderate,True,False,False,False,False,True
sqft_living15,price,Spearman,0.572,0.000,Moderate,True,False,False,False,False,True
sqft_above,price,Spearman,0.542,0.000,Moderate,True,False,False,False,True,True
bathrooms,price,Spearman,0.497,0.000,Moderate,True,False,False,False,False,True
bedrooms,price,Spearman,0.345,0.000,Moderate,True,False,False,False,False,True
floors,price,Spearman,0.322,0.000,Moderate,True,False,False,False,False,True
sqft_basement,price,Spearman,0.252,0.000,Weak,True,False,False,False,True,True
yr_built,price,Spearman,0.102,0.000,Weak,True,False,False,False,False,True


[ 상관 강도 판정별 변수 개수 ]
strength
Moderate    7
Weak        4
Name: count, dtype: int64


#### 이변량 분석 - T검정 (2집단 명목형 → 종속변수)

In [9]:
for n in nominal_cols:
    # 2개의 고유값만 존재하는 범주형 독립변수에 대해서만 검정을 수행한다.
    if len(df[n].unique()) == 2:
        display(Markdown(f"#### ▶︎ {n}({column_means[n]}) → {target}"))

        # 수집된 데이터가 long 타입이므로 검정 대상만 wide 타입으로 변환해야 한다.
        wide_df = my_prep.long2wide(df, hue=n, values=target)
        display(my_stats.test_independent(wide_df,
                    group1=wide_df.columns[0], group2=wide_df.columns[1], plot=False))

        display(df.groupby(n, observed=True)[target].agg(['count', 'mean']))

#### ▶︎ waterfront(워터프론트 여부) → price

statistic  p-value  significant result
test                alternative                                        
Mann-Whitney U test two-sided   405519.000    0.000         True  0 ≠ 1
                    less        405519.000    0.000         True  0 < 1
                    greater     405519.000    1.000        False  0 ≤ 1

,count,mean
waterfront,,
0,21450,531563.600
1,163,1661876.025


#### 이변량 분석 - ANOVA (3집단 이상 명목형 → 종속변수)

In [10]:
for n in nominal_cols:
    # 3개 이상의 고유값이 존재하는 명목형 독립변수에 대해서만 검정을 수행한다.
    if len(df[n].unique()) >= 3:
        display(Markdown(f"#### ▶︎ {n}({column_means[n]}) → {target}"))

        display(my_stats.anova_oneway(df, y=target, between=n))

        # 사후검정
        display(my_stats.posthoc_oneway(df, y=target, between=n, plot=False))

#### ▶︎ view(조망 점수) → price

,test,Source,ddof1,ddof2,F,p_unc,np2,effect_size
0,welch_anova,view,4,881.442,264.759,0.000,0.168,Large


비교한 조합 쌍: 10개
유의한 쌍: 9개
평균 차이가 가장 큰 쌍: (0) vs (4) → 차이 967147.046, 효과크기 Large, p-value 0.000
평균 차이가 가장 작은 쌍: (1) vs (2) → 차이 19879.949, 효과크기 Negligible, p-value 0.973


,test,A,B,mean_A,mean_B,diff,se,T,df,pval,hedges,significant,effect_size
0,Games-Howell,0,1,496564.192,812280.837,-315716.645,28117.335,-11.229,334.571,0.000,-1.080,True,Large
1,Games-Howell,0,2,496564.192,792400.888,-295836.696,16566.083,-17.858,992.346,0.000,-0.982,True,Large
2,Games-Howell,0,3,496564.192,971965.275,-475401.082,27208.320,-17.473,514.867,0.000,-1.585,True,Large
3,Games-Howell,0,4,496564.192,1463711.238,-967147.046,53353.159,-18.127,318.947,0.000,-3.127,True,Large
4,Games-Howell,1,2,812280.837,792400.888,19879.949,32504.752,0.612,574.228,0.973,0.039,False,Negligible
5,Games-Howell,1,3,812280.837,971965.275,-159684.437,39018.156,-4.093,790.353,0.000,-0.278,True,Small
6,Games-Howell,1,4,812280.837,1463711.238,-651430.401,60238.554,-10.814,482.793,0.000,-0.856,True,Large
7,Games-Howell,2,3,792400.888,971965.275,-179564.387,31721.713,-5.661,887.980,0.000,-0.328,True,Small
8,Games-Howell,2,4,792400.888,1463711.238,-671310.350,55790.089,-12.033,380.199,0.000,-1.034,True,Large
9,Games-Howell,3,4,971965.275,1463711.238,-491745.964,59819.659,-8.220,483.759,0.000,-0.645,True,Medium


#### ▶︎ condition(주택 상태 점수) → price

,test,Source,ddof1,ddof2,F,p_unc,np2,effect_size
0,welch_anova,condition,4,197.161,52.777,0.000,0.007,Negligible


비교한 조합 쌍: 10개
유의한 쌍: 9개
평균 차이가 가장 큰 쌍: (2) vs (5) → 차이 285130.944, 효과크기 Medium, p-value 0.000
평균 차이가 가장 작은 쌍: (1) vs (2) → 차이 7144.521, 효과크기 Negligible, p-value 1.000


,test,A,B,mean_A,mean_B,diff,se,T,df,pval,hedges,significant,effect_size
0,Games-Howell,1,2,334431.667,327287.145,7144.521,52927.610,0.135,37.747,1.000,0.029,False,Negligible
1,Games-Howell,1,3,334431.667,542012.578,-207580.911,49604.665,-4.185,29.224,0.002,-0.570,True,Medium
2,Games-Howell,1,4,334431.667,521200.390,-186768.723,49737.205,-3.755,29.538,0.006,-0.521,True,Medium
3,Games-Howell,1,5,334431.667,612418.089,-277986.423,50501.977,-5.504,31.396,0.000,-0.679,True,Medium
4,Games-Howell,2,3,327287.145,542012.578,-214725.433,18964.229,-11.323,180.369,0.000,-0.591,True,Medium
5,Games-Howell,2,4,327287.145,521200.390,-193913.245,19308.256,-10.043,193.795,0.000,-0.545,True,Medium
6,Games-Howell,2,5,327287.145,612418.089,-285130.944,21200.681,-13.449,279.464,0.000,-0.715,True,Medium
7,Games-Howell,3,4,542012.578,521200.390,20812.188,5665.650,3.673,10665.855,0.002,0.057,True,Negligible
8,Games-Howell,3,5,542012.578,612418.089,-70405.511,10428.793,-6.751,2037.356,0.000,-0.190,True,Negligible
9,Games-Howell,4,5,521200.390,612418.089,-91217.699,11042.033,-8.261,2524.069,0.000,-0.246,True,Small


#### 다변량 분석 - 독립변수간 상관 분석

In [11]:
corr = my_stats.multi_correlation(df, columns=continuous_cols, plot=False)
my_stats.correlation_summary(corr)

,bedrooms,bathrooms,sqft_living,sqft_lot,floors,grade,sqft_above,sqft_basement,yr_built,sqft_living15,sqft_lot15
bedrooms,1.000,0.521,0.647,0.216,0.228,0.381,0.540,0.231,0.180,0.444,0.202
bathrooms,0.521,1.000,0.746,0.069,0.547,0.658,0.691,0.192,0.567,0.570,0.063
sqft_living,0.647,0.746,1.000,0.304,0.401,0.716,0.844,0.328,0.352,0.747,0.284
sqft_lot,0.216,0.069,0.304,1.000,-0.234,0.152,0.272,0.037,-0.038,0.360,0.922
floors,0.228,0.547,0.401,-0.234,1.000,0.502,0.599,-0.273,0.552,0.305,-0.231
grade,0.381,0.658,0.716,0.152,0.502,1.000,0.712,0.093,0.501,0.663,0.156
sqft_above,0.540,0.691,0.844,0.272,0.599,0.712,1.000,-0.166,0.472,0.697,0.254
sqft_basement,0.231,0.192,0.328,0.037,-0.273,0.093,-0.166,1.000,-0.178,0.130,0.031
yr_built,0.180,0.567,0.352,-0.038,0.552,0.501,0.472,-0.178,1.000,0.336,-0.016
sqft_living15,0.444,0.570,0.747,0.360,0.305,0.663,0.697,0.130,0.336,1.000,0.366


,max-y,max-coef,count,columns
x,,,,
sqft_living,sqft_above,0.844,4,"sqft_above, sqft_living15, bathrooms, grade"
sqft_above,sqft_living,0.844,2,"sqft_living, grade"
grade,sqft_living,0.716,2,"sqft_living, sqft_above"
sqft_lot,sqft_lot15,0.922,1,sqft_lot15
sqft_lot15,sqft_lot,0.922,1,sqft_lot
sqft_living15,sqft_living,0.747,1,sqft_living
bathrooms,sqft_living,0.746,1,sqft_living
bedrooms,sqft_living,0.647,0,-
floors,sqft_above,0.599,0,-


### 문제 풀이

#### 1. 캘리포니아에서는 조사 상한에 값이 몰린 탓에 가격 분포가 오른쪽으로 치우쳤습니다. 킹 카운티 가격에는 그런 절단 지점이 없는데도 분포는 여전히 오른쪽으로 길게 늘어져 있습니다. 종속변수의 왜도는 얼마인가요? (소수 셋째 자리)

- **정답**: `4.024`
- **복습 개념**: 연속형 종속변수의 단변량 분석입니다. 분포 모양을 왜도로, 뾰족한 정도를 첨도로 읽고, 그 값이 로그 변환 필요 여부 판정으로 이어집니다.
- **풀이 접근**: 종속변수와 연속형 독립변수를 한 데 넣어 기술통계량 표를 만든 뒤, 종속변수 행의 왜도 열을 읽습니다. 같은 행의 첨도·이상치 비율·로그 필요 여부도 함께 보면 근거가 겹쳐 쌓입니다.
- **근거(계산 결과)**: 가격의 왜도는 **4.024**, 첨도는 34.586입니다. 캘리포니아 주택 가격의 왜도가 +0.978이었으니 **네 배가 넘습니다.** 평균 540,088달러가 중앙값 450,000달러보다 위에 있고, 최댓값은 7,700,000달러로 중앙값의 17배입니다. 로그 필요 여부 판정도 `log`로 찍힙니다.
- **헷갈리기 쉬운 점**: 캘리포니아의 우편향은 500,001달러에 값이 인위적으로 쌓여서 생긴 것이라 "절단을 걷어내면 완화된다"고 볼 여지가 있었습니다. 킹 카운티에는 그런 한 점이 없습니다. 여기서 꼬리를 만드는 것은 실제로 존재하는 770만 달러짜리 저택들입니다. **같은 우편향이라도 원인이 다르면 조치도 달라집니다.** 이쪽은 지울 값이 아니라 로그 변환으로 다룰 값입니다.

#### 2. 캘리포니아에서는 연속형 독립변수 6종 가운데 상관 강도가 '보통' 이상으로 판정된 것이 중위 소득 하나뿐이었습니다. 이 자료의 연속형 독립변수 11종 중 '보통'으로 판정된 변수는 몇 개인가요? (정수, 단위: 개)

- **정답**: `7`
- **복습 개념**: 상관분석 결과표의 강도 판정입니다. 상관계수의 절댓값이 0.7을 넘으면 강함, 0.3 초과 0.7 이하면 보통, 0 초과 0.3 이하면 약함으로 라벨이 붙습니다.
- **풀이 접근**: 연속형 독립변수 이름을 하나씩 돌면서 종속변수와의 상관분석을 수행하고, 결과를 한 표로 합칩니다. 그 표의 강도 열만 뽑아 값별로 개수를 셉니다. 참고로 11종이라는 숫자는 지난 단원에서 값의 범위를 점검했던 15개에서 종속변수(price)와 이번에 범주형으로 돌린 두 종(view·condition), 그리고 제외한 리모델링 연도를 뺀 결과입니다.
- **근거(계산 결과)**: 판정 결과는 `Moderate` **7개**(grade 0.658 / sqft_living 0.644 / sqft_living15 0.572 / sqft_above 0.542 / bathrooms 0.497 / bedrooms 0.345 / floors 0.322), `Weak` 4개(sqft_basement 0.252 / yr_built 0.102 / sqft_lot 0.075 / sqft_lot15 0.063)입니다. `Strong`은 0개이며, 11종 모두 p가 0.000이라 유의성 판정은 전부 True입니다.
- **함께 생각해 볼 점**: 캘리포니아는 구역 단위로 집계된 자료라 집 한 채의 특성이 평균에 묻혀 상관이 잘 잡히지 않았습니다. 이 자료는 주택 한 채가 한 행이라 면적·등급·욕실 수가 가격과 곧바로 맞물립니다. **설명변수가 많아진 것이 아니라 집계 단위가 달라진 것입니다.** 다만 쓸 만한 변수가 7개라는 뜻은 아닙니다. 이 7종이 서로 겹치는지는 다변량 단계에서 따로 확인해야 합니다.

#### 3. 캘리포니아 EDA 범위표에서 대상이 없어 '미적용'으로 비어 있던 두 집단 비교를, 이번에는 워터프론트 여부를 두고 수행하게 됩니다. 정규성 점검 결과에 따라 채택된 검정의 이름은 무엇인가요? (검정 이름 1개)

- **정답**: `Mann-Whitney U test`
- **복습 개념**: 두 독립집단 비교에서의 검정 자동 선택입니다. 두 집단이 모두 정규성을 충족하면 등분산 여부에 따라 Student 또는 Welch t검정을, 하나라도 충족하지 못하면 비모수 검정을 씁니다.
- **풀이 접근**: 명목형 변수 중 고유값이 2개인 것만 골라내고, 집단별 가격을 열로 나눈 형태로 바꾼 뒤 두 집단 비교를 수행합니다. 결과표의 검정 이름 열을 읽으면 어떤 절차가 선택되었는지 그대로 나옵니다.
- **근거(계산 결과)**: 결과표에 **Mann-Whitney U test**로 찍힙니다. 가격의 왜도가 4.024여서 두 집단 모두 정규성을 충족하지 못했기 때문입니다. 양측검정 p는 0.000으로 유의하고, 워터프론트인 주택 163채의 평균은 1,661,876달러로 나머지 21,450채의 531,564달러보다 **3배 이상** 높습니다.
- **실무 포인트**: "두 집단 비교 = t검정"으로 외워 두면 여기서 어긋납니다. 검정 이름은 데이터의 가정 점검 결과가 정하는 것이지 집단 수가 정하는 것이 아닙니다. 상관분석에서 Pearson 대신 Spearman이 선택된 것과 같은 원리입니다. 다만 이 결과를 그대로 믿기 전에 표본을 보세요. 워터프론트 주택은 **163채(0.8%)** 뿐입니다. 캘리포니아의 ISLAND 5건만큼 극단적이지는 않지만, 더미 인코딩 후 계수 해석에는 주의가 필요합니다.

#### 4. 3집단 이상인 명목형 변수 두 종에 대해 집단 간 평균 비교를 수행했더니 둘 다 유의확률이 0.001보다 작게 나왔습니다. 그런데 그중 하나는 효과크기가 '무시할 수준'으로 판정되었습니다. 그 변수는 무엇인가요? (변수명 1개)

- **정답**: `condition`
- **복습 개념**: 유의성과 효과크기의 구분입니다. 유의확률은 "차이가 있느냐"에, 효과크기(np2)는 "그 차이가 얼마나 크냐"에 답합니다. 표본이 크면 아주 작은 차이도 유의하게 나옵니다.
- **풀이 접근**: 고유값이 3개 이상인 명목형 변수만 골라 집단 간 평균 비교를 수행하고, 결과표의 효과크기 열을 서로 비교합니다. 이어서 사후검정으로 어떤 쌍이 실제로 갈리는지 확인합니다.
- **근거(계산 결과)**: **condition**은 F=52.777로 p가 0.000이지만 np2=**0.007**이라 판정이 `Negligible`입니다. 반면 view는 np2=0.168로 `Large`입니다. condition의 집단 평균을 보면 3등급 542,013달러, 4등급 521,200달러로 **등급이 높은 쪽이 오히려 싸고**, 사후검정에서도 3 대 4 쌍의 효과크기가 Negligible입니다. 21,613건이라는 표본 크기가 2만 달러 차이를 유의하게 만든 것뿐입니다.
- **자주 하는 실수**: p가 0.000이라는 이유만으로 채택 목록에 올리기 쉽습니다. 캘리포니아의 해안권 3종이 "유의하지만 효과크기 Negligible"이라 병합 대상이 되었던 것과 같은 상황입니다. **표본이 크면 p는 거의 항상 작아지므로, 채택 여부는 효과크기로 판단하세요.** 반대로 view는 조망 0점 496,564달러 대 4점 1,463,711달러로 100만 달러 가까이 벌어지는 진짜 설명변수입니다.

#### 5. 이제 팀장에게 1순위 변수를 보고해야 합니다. 종속변수와의 상관이 가장 강했고, 독립변수끼리 겹치는 덩어리 안에서도 대표로 남길 만한 변수는 무엇인가요? (변수명 1개)

- **정답**: `grade`
- **복습 개념**: 다변량 단계의 중복 점검과 최종 변수 선택입니다. 이변량에서 살아남은 변수라도 서로 강하게 겹치면 하나만 남기고, 그때 기준은 종속변수와의 설명력입니다.
- **풀이 접근**: 연속형 독립변수 전체의 상관행렬을 만들고, 강한 상관으로 묶인 변수 덩어리를 요약합니다. 그 덩어리 안에서 이변량 결과표의 상관계수를 비교해 대표를 고릅니다.
- **근거(계산 결과)**: **grade**의 상관계수가 0.658로 11종 중 1위입니다. 중복 점검에서 grade는 sqft_living(0.716)·sqft_above와 한 덩어리로 묶이는데, 그 덩어리에서 상관계수가 가장 높은 쪽이 grade(0.658 > sqft_living 0.644 > sqft_above 0.542)입니다. 규모 축의 허브인 sqft_living은 무려 4개 변수(sqft_above 0.844 / sqft_living15 0.747 / bathrooms 0.746 / grade 0.716)와 강하게 겹칩니다.
- **결론**: 팀장에게 드릴 답은 이렇게 정리됩니다. **1순위는 `grade`, 2순위는 명목형에서 효과크기가 가장 큰 `view`입니다.** 규모 축(sqft_living·sqft_above·sqft_living15·bathrooms)은 서로 0.7 이상으로 겹치므로 grade 하나로 대표하고, sqft_lot과 sqft_lot15는 서로 0.922로 겹치면서 가격과의 상관은 0.07 수준이라 함께 제외합니다. condition은 유의하지만 효과크기가 없어 보류입니다. 캘리포니아에서는 위치(ocean_proximity)와 소득이 답이었는데, 킹 카운티에서는 **건축 등급과 조망**이 답입니다. 절차는 똑같았지만 절차가 가리킨 변수는 데이터마다 달랐습니다.